### PR TITLE
[PeerConnectors] Fix connecting to multiple nodes when -addnode and -connect args are specfied and one peer fails.

### LINKED ISSUE #967: [PeerConnectors] AddNode and Connect doesn't connect to all peers if one fails.
Currently, if you specify 2 nodes using -addnode and try and connect to the first peer and it fails, we dont continue onto the next one. This applies to -connect as well.

A slight refactoring needs to happen as there is a design flaw on how PeerConnectorAddNode and PeerConnectorConnect connects to peers.

### PR DESCRIPTION
**Pull request by fassadlr**
This fixes #967 .

### CODE REVIEW

File: src/Stratis.Bitcoin/P2P/PeerConnector.cs [hunk 1]
```diff
@@ -191,27 +184,22 @@ public void StartConnectAsync()
...
-            if (this.ConnectedPeers.Count >= this.MaximumNodeConnections)
-                return;
-
-            PeerAddress peerAddress = this.FindPeerToConnectTo();
-            if (peerAddress == null)
-            {
-                Task.Delay(TimeSpans.TenSeconds.Milliseconds).Wait(this.nodeLifetime.ApplicationStopping);
-                return;
-            }
+            this.logger.LogTrace("()");
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/PeerConnector.cs]: you are missing matching exit log
and you should also log the argument

File: src/Stratis.Bitcoin/P2P/PeerConnector.cs [hunk 2]
```diff
@@ -191,27 +184,22 @@ public void StartConnectAsync()
...
             this.asyncLoop = this.asyncLoopFactory.Run($"{this.GetType().Name}.{nameof(this.ConnectAsync)}", async token =>
             {
-                await this.ConnectAsync();
+                if (!this.peerAddressManager.Peers.Any())
+                    return;
+
+                if (this.ConnectedPeers.Count >= this.MaximumNodeConnections)
+                    return;
+
+                await this.OnConnectAsync();
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/PeerConnector.cs]: according to gentleman agreement made earlier somewhere, we do use .ConfigureAwait(false) with all awaits in all FN code (not tests though)

File: src/Stratis.Bitcoin/P2P/PeerConnectorAddNode.cs [hunk 3]
```diff
@@ -59,18 +60,16 @@ public override void OnStartConnect()
...
         /// </summary>
-        public override PeerAddress FindPeerToConnectTo()
+        public async override Task OnConnectAsync()
         {
             foreach (var ipEndpoint in this.NodeSettings.ConnectionManager.AddNode)
             {
                 PeerAddress peerAddress = this.peerAddressManager.FindPeer(ipEndpoint);
                 if (peerAddress != null && !this.IsPeerConnected(peerAddress.NetworkAddress.Endpoint))
-                    return peerAddress;
+                    await ConnectAsync(peerAddress);
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/PeerConnectorAddNode.cs]: configure await here as well

File: src/Stratis.Bitcoin/P2P/PeerConnectorConnect.cs [hunk 4]
```diff
@@ -60,18 +61,16 @@ public override void OnStartConnect()
...
         /// </summary>
-        public override PeerAddress FindPeerToConnectTo()
+        public async override Task OnConnectAsync()
         {
             foreach (var ipEndpoint in this.NodeSettings.ConnectionManager.Connect)
             {
                 PeerAddress peerAddress = this.peerAddressManager.FindPeer(ipEndpoint);
                 if (peerAddress != null && !this.IsPeerConnected(peerAddress.NetworkAddress.Endpoint))
-                    return peerAddress;
+                    await ConnectAsync(peerAddress);
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/PeerConnectorConnect.cs]: and here

File: src/Stratis.Bitcoin/P2P/PeerConnectorConnect.cs [hunk 5]
```diff
@@ -60,18 +61,16 @@ public override void OnStartConnect()
         }
 
         /// <summary>
-        /// Only return nodes as specified in the -connect node arg.
+        /// Only connect to nodes as specified in the -connect node arg.
         /// </summary>
-        public override PeerAddress FindPeerToConnectTo()
+        public async override Task OnConnectAsync()
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/PeerConnectorConnect.cs]: elsewhere in the code we use `public override async` in that order, let's be consistent

File: src/Stratis.Bitcoin/P2P/PeerConnectorDiscovery.cs [hunk 6]
```diff
@@ -58,8 +59,7 @@ public override void OnStartConnect()
             this.CurrentParameters.PeerAddressManagerBehaviour().Mode = PeerAddressManagerBehaviourMode.AdvertiseDiscover;
         }
 
-        /// <inheritdoc/>
-        public override PeerAddress FindPeerToConnectTo()
+        public async override Task OnConnectAsync()
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/PeerConnectorDiscovery.cs]: here as well

File: src/Stratis.Bitcoin/P2P/PeerConnectorAddNode.cs [hunk 7]
```diff
@@ -59,18 +60,16 @@ public override void OnStartConnect()
         }
 
         /// <summary>
-        /// Only return nodes as specified in the -addnode arg.
+        /// Only connect to nodes as specified in the -addnode arg.
         /// </summary>
-        public override PeerAddress FindPeerToConnectTo()
+        public async override Task OnConnectAsync()
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/PeerConnectorAddNode.cs]: and here

File: src/Stratis.Bitcoin/P2P/PeerConnectorDiscovery.cs [hunk 8]
```diff
@@ -124,8 +124,12 @@ public override PeerAddress FindPeerToConnectTo()
                 break;
             }
 
-            this.logger.LogTrace("(-):'{0}'", peer?.NetworkAddress.Endpoint);
-            return peer;
+            this.logger.LogTrace("Peer selected: '{0}'", peer?.NetworkAddress.Endpoint);
+
+            if (peer != null)
+                await ConnectAsync(peer);
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/P2P/PeerConnectorDiscovery.cs]: and configure await

File: src/Stratis.Bitcoin/P2P/PeerConnector.cs (hunk 1 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerConnector.cs]: Fixed

File: src/Stratis.Bitcoin/P2P/PeerConnector.cs (hunk 2 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerConnector.cs]: Done 👍

File: src/Stratis.Bitcoin/P2P/PeerConnectorAddNode.cs (hunk 3 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerConnectorAddNode.cs]: Done 👍

File: src/Stratis.Bitcoin/P2P/PeerConnectorConnect.cs (hunk 4 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerConnectorConnect.cs]: Done 👍

File: src/Stratis.Bitcoin/P2P/PeerConnectorConnect.cs (hunk 5 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerConnectorConnect.cs]: Done 👍

File: src/Stratis.Bitcoin/P2P/PeerConnectorDiscovery.cs (hunk 6 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerConnectorDiscovery.cs]: Done 👍

File: src/Stratis.Bitcoin/P2P/PeerConnectorAddNode.cs (hunk 7 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerConnectorAddNode.cs]: Done 👍

File: src/Stratis.Bitcoin/P2P/PeerConnectorDiscovery.cs (hunk 8 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin/P2P/PeerConnectorDiscovery.cs]: Done 👍

**Pull request merged**